skimage/morphology/_max_tree.pyx

```diff
@@ -0,0 +1,459 @@
+                         DTYPE_INT64_t[::1] parent,
+                         DTYPE_INT64_t[::1] sorted_indices
+                         ):
+    """Finds the local maxima in image.
+
+    Parameters
+    ----------
+
+    image : array of arbitrary type
+        The flattened image pixels.
+    output : array of the same shape and type as image.
+        The output image must contain only ones.
+    label : bool.
+        Indicates whether the minima are to be labeled.
+    parent : array of int
+        Image of the same shape as the input image. The value
+        at each pixel is the parent index of this pixel in the max-tree
+        reprentation.
+    sorted_indices : array of int
+        List of length = number of pixels. Each element
+        corresponds to one pixel index in the image. It encodes the order
+        of elements in the tree: a parent of a pixel always comes before
+        the element itself. More formally: i < j implies that j cannot be
+        the parent of i.
+    """
+
+    cdef DTYPE_INT64_t p_root = sorted_indices[0]
+    cdef DTYPE_INT64_t p, q
+    cdef DTYPE_UINT64_t number_of_pixels = len(image)
+    cdef DTYPE_UINT64_t label = 1
+
+    for p in sorted_indices[::-1]:
+        if p == p_root:
+            continue
+
+        q = parent[p]
+
+        # if p is canonical (parent has a different value)
+        if image[p] != image[q]:
+            output[q] = 0
+            if label_minima:
+                # if output[p] was the parent of some other canonical
+                # pixel, it has been set to zero. Only the leaves
+                # (local maxima) are thus > 0.
+                if output[p] > 0:
+                    output[p] = label
+                    label += 1
+            continue
+
+    for p in sorted_indices[::-1]:
+        if p == p_root:
+            continue
+
+        q = parent[p]
+
+        # if p is not canonical (parent has the same value)
+        if image[p] == image[q]:
+            # in this case we propagate the value
+            output[p] = output[q]
+            continue
+
+    return
+
+# _cut_first_filter is similar to _direct_filter, but it stops the
+# reconstruction process when a criterion is fulfilled for the first time.
+# For increasing criteria this is equivalent to _direct_filter.
+cpdef void _cut_first_filter(dtype_t[::1] image,
+                             dtype_t[::1] output,
+                             DTYPE_INT64_t[::1] parent,
+                             DTYPE_INT64_t[::1] sorted_indices,
+                             DTYPE_FLOAT64_t[::1] attribute,
+                             DTYPE_FLOAT64_t attribute_threshold
+                             ):
+    """Image filter using max-trees: starting from the root, the tree is pruned
+    if the attribute of a component is smaller than the threshold. For
+    increasing attributes, this is the same as _direct_filter, but for
+    non-increasing attributes, the results are different (and more severe).
```

Comment:
What does "more severe" mean? A gallery example of this might also be useful...

Reply:
Here severe means that it modifies the image more. 

Actually, I would like to postpone this to a new PR, when I will include a non-increasing attribute. At the moment, I do not have time to do this, unfortunately. I think that if I give a concrete example, this is getting much clearer. At the moment, I agree this remains enigmatic. We can also remove this function from the PR altogether (as it is useless for increasing criteria). As you wish ...

Reply:
That's fine. Let's only keep in this PR what is necessary, and simple enough that even @stefanv or I can fully understand it. =) I think I need to read the source paper to provide more specific advice... Probably won't have time to do this before SciPy though.

Reply:
Come on, if you don't understand this, nobody will. And actually, it is quite logical that this is not really understandable in the absence of a concrete operator ... so, yes, I will remove the function from this PR and propose it again, when I add the elliptical filters in a new PR.

Reply:
It will be much clearer then.